src/ucp/tag/rndv.c

```diff
@@ -44,26 +69,86 @@ static void ucp_rndv_rma_request_send_buffer_dereg(ucp_request_t *sreq)
 {
-    ucp_ep_h ep = sreq->send.ep;
-    ucp_lane_index_t lane = ucp_ep_get_rndv_get_lane(ep, 0);
+    ucp_ep_h ep           = sreq->send.ep;
+    uint8_t *buf          = rkey_buf;
+    uint8_t *cnt          = buf;
+    size_t packet;
+    ucp_lane_index_t lane;
     ucs_status_t status;
+    int i;
+    ucp_rndv_rkey_data_t *rkey;
 
     ucs_assert(UCP_DT_IS_CONTIG(sreq->send.datatype));
 
-    /* Check if the sender needs to register the send buffer -
-     * is its datatype contiguous and does the receive side need it */
-    if (ucp_ep_rndv_md_flags(ep, 0) & UCT_MD_FLAG_NEED_RKEY) {
-        status = ucp_request_rndv_buffer_reg(sreq);
-        ucs_assert_always(status == UCS_OK);
-
-        /* if the send buffer was registered, send the rkey */
-        UCS_PROFILE_CALL(uct_md_mkey_pack, ucp_ep_md(ep, lane),
-                         sreq->send.state.dt.dt.contig[0].memh, rkey_buf);
+    buf++; /* first byte is counter of packet keys */
+    *cnt   = 0;
+    packet = 1; /* packet at least counter */
+
+    status = ucp_request_rndv_buffer_reg(sreq);
+    ucs_assert_always(status == UCS_OK);
+
+    for (i = 0; i < ucp_ep_rndv_num_lanes(ep); i++) {
+        lane = ucp_ep_get_rndv_get_lane(ep, i);
+        rkey = (ucp_rndv_rkey_data_t*)(buf + packet);
+        rkey->md_index = ucp_ep_dst_md_index(ep, lane);
+        /* Check if the sender needs to register the send buffer -
+         * is its datatype contiguous and does the receive side need it */
+        if (ucp_ep_rndv_md_flags(ep, i) & UCT_MD_FLAG_NEED_RKEY) {
+            rkey->key_size = (uint8_t)ucp_ep_md_attr(ep, lane)->rkey_packed_size;
+
+            /* if the send buffer was registered, send the rkey */
+            UCS_PROFILE_CALL(uct_md_mkey_pack, ucp_ep_md(ep, lane),
+                             sreq->send.state.dt.dt.contig[i].memh, rkey->rkey);
+        } else {
+            /* in case if remote key is not used - just let peer to know about MD used */
+            rkey->key_size = 0;
+        }
         *flags |= UCP_RNDV_RTS_FLAG_PACKED_RKEY;
-        return ucp_ep_md_attr(ep, lane)->rkey_packed_size;
+        packet += sizeof(*rkey) + rkey->key_size;
     }
 
-    return 0;
+    ucs_assert_always(packet <= ucp_ep_config(ep)->am.max_bcopy);
+    return ((*flags) & UCP_RNDV_RTS_FLAG_PACKED_RKEY) ? packet : 0;
+}
+
+static void ucp_tag_rndv_unpack_send_rkey(ucp_request_t *req, void *rkey_buf)
+{
+    /* rkey buffer has format:
+     * |count 1 byte|md_index 1 byte|key size 1 byte|remote key|md_index|key size|remote key|... */
+
+    ucp_ep_h ep  = req->send.ep;
+    uint8_t *buf = rkey_buf;
+    uint8_t  cnt = *buf;
+    ucp_lane_index_t lane;
+    int i;
+    ucp_rndv_rkey_data_t *rkey;
+
+    ucs_assert_always(cnt < UCP_MAX_RNDV_LANES);
+    ucs_assert(req->send.rndv_get.rkey != NULL);
+
+    buf++; /* first byte is counter of packet keys */
+    req->send.rndv_get.num_lanes = 0;
+
+    for (i = 0; i < cnt; i++) {
```

Comment:
datatype mismatch between `i` and `cnt`, it's probably reason of ORNL failure. also minor - it's better to reorder variables on stack from big to small.

Reply:
> datatype mismatch between i and cnt, it's probably reason of ORNL failure
nop, there is another issue, fix is in progress
> also minor - it's better to reorder variables on stack from big to small
really strange rule :)
actually most of the variables will not be located at stack (have you seen debug message about variable value is optimized?), or will be located in order which is most performance gain

Reply:
`int i;`
...
`uint8_t  cnt = *buf;`

> really strange rule

ok, it's probably out-of-date rule and my personal preference :)